### DISCUSSION TITLE
Inverted RX serial in Berry

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#20706] [Q&A] [answered]
I want to read an inverted serial signal from a Dutch smartmeter with Tasmota using this device:

https://smart-stuff.nl/product/p1-dongel-slimme-meter-esp32/

The serial data is inverted by using the espressif function uartSetRxInvert.

How can I set Tasmota to receive the inverted serial data in Berry without additional hardware?

**Comment by arendst** [comment]
I guess this would need a change in Berry ``b_serial_init`` supporting the recently added invert option to TasmotaSerial()

**Comment by s-hadinger** [comment]
Let me add this option, it's simple enough.

**Comment by s-hadinger** [comment, marked as answer]
Please use the latest development version, there is now an optional argument to set to inverted:

```berry
# gpio_rx:4 gpio_tx:5

# normal mode
ser = serial(4, 5, 9600, serial.SERIAL_7E1)

# inverted mode
ser = serial(4, 5, 9600, serial.SERIAL_7E1, true)
```

**Reply by homeplayer2** [reply]
> Please use the latest development version, there is now an optional argument to set to inverted:
> 
> ```berry
> # gpio_rx:4 gpio_tx:5
> 
> # normal mode
> ser = serial(4, 5, 9600, serial.SERIAL_7E1)
> 
> # inverted mode
> ser = serial(4, 5, 9600, serial.SERIAL_7E1, true)
> ```

its 4 years later, sorry for the bump,
but is it possible now to set 115200, SERIAL_8N1, 16, -1, true
aka rx 16 tx null signal inverted
or do i need a custom build, np there but where and how to set it?

**Reply by s-hadinger** [reply]
Those are in the standard builds. It should work. Just try it.

**Comment by jeroenst** [comment]
Thank you very much!

I'm going to test it this week.

**Comment by georgewahba** [comment]
hey man, I have the same dongle but I am not getting it to work, did you figure it out? so yes are you able to help me with mine?

**Comment by jeroenst** [comment]
Not yet, I want to pick up the development of the smartmeter script in the next few weeks
